maze_transformer/evaluation/path_evals.py

```diff
@@ -0,0 +1,109 @@
+        yield (n_s, n_e)
+
+
+class PathEvals:
+    """array path based eval functions. first path is always the "ground truth" path"""
+
+    @staticmethod
+    def node_overlap(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """number of shared nodes (any order) / total number of (unique) nodes"""
+        if len(prediction) <= 1:
+            return 0.0
+
+        n_shared: int = 0
+        solution_set = {tuple(coord) for coord in solution}
+        prediction_set = {tuple(coord) for coord in prediction}
+
+        for coord in solution:
+            if tuple(coord) in prediction_set:
+                n_shared += 1
+        return n_shared / len(solution_set)
+
+    @staticmethod
+    def num_connections_adjacent_lattice(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """number of the connections in prediction which actually connect nodes that are adjacent on the lattice, ignoring if they are adjacent on the maze"""
+        if len(prediction) <= 1:
+            return 0.0
+
+        n_adj: int = 0
+        for n_s, n_e in path_as_segments_iter(prediction):
+            # print(f"{n_s = } {n_e = }")
+
+            if (np.abs(n_s - n_e).sum() == 1).all():
+                n_adj += 1
+
+        return n_adj
+
+    @staticmethod
+    def fraction_connections_adjacent_lattice(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """fraction of the connections in prediction which actually connect nodes that are adjacent on the lattice, ignoring if they are adjacent on the maze"""
+
+        return PathEvals.num_connections_adjacent_lattice(
+            maze, solution, prediction
+        ) / len(prediction)
+
+    @staticmethod
+    def num_connections_adjacent(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """number of connections in prediction which are are valid paths on the maze"""
+
+        if len(prediction) <= 1:
+            return 0.0
+
+        n_connected: int = 0
+        for n_s, n_e in path_as_segments_iter(prediction):
+            if maze.nodes_connected(n_s, n_e):
+                n_connected += 1
+
+        return n_connected
+
+    @staticmethod
+    def fraction_connections_adjacent(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
+    ) -> float:
+        """fraction of connections in prediction which are are valid paths on the maze"""
+
+        return PathEvals.num_connections_adjacent(maze, solution, prediction) / len(
+            prediction
+        )
+
+    @classmethod
+    def all_functions(cls) -> dict[str, PathEvalFunction]:
+        excluded = ["all_functions"]
+
+        return {
+            **{
+                name: func
+                for name, func in cls.__dict__.items()
+                if not name.startswith("_") and name not in excluded
+            }
+        }
```

Comment:
Yep this is a fair concern. Enforcing types more strongly would help for sure. 

If there's a more elegant solution I don't know what it is - I did think this code was a bit concerning when I moved it in here. But there is a valid need here for us to be able to easily get a list of all path eval functions. I didn't want to just add some constant listing them somewhere because then there would be 2 places to update any time we change something. 

If there's a more Pythonic way of doing this I'm definitely open to it!

Reply:
I wonder if I could add some checks to this function to make sure everything being returned adheres to the correct interface.

Reply:
Adding the following checks might help:
- `isinstance(func, typing.Callable)`
- `isinstance(func, types.FunctionType)` to [check it's a function type](https://stackoverflow.com/questions/8727059/python-check-if-method-is-static)
- `typing.get_type_hints(func)["return"] is float`

`typing.get_type_hints(func)` can also help us check if the argument names & types are correct

a few other notes:
- we probably want to wrap the calls to these functions in a `try ... catch` since we dont want the whole training script to crash if an eval fails. definitely print warnings though, and have a way to make it crash during integration tests. maybe this is already being done?
- is there a reason we are unwrapping the dict only to wrap it again? (maybe this is my fault haha)
- do we want to perform some type checks when loading the module, rather than waiting for `all_functions` to be called elsewhere? this should definitely throw an exception if it finds any functions which dont fit the criteria.
- should we rename `all_functions` to `_all_functions`, thereby excluding it automatically (or maybe move out of the class)? hopefully this discourages adding non-evaluation functions to this class

Reply:
I agree with Michael that the dictionary seems to be being unnecessarily unpacked and repacked. That this syntax slipped through is a sign that no one understands the syntax, which means **we shouldn't use that syntax**.

We should aim for code that is readable and maintainable rather than code which is clever. In this case I think you need to update the eval functions to remove unused parameters. Then you need to update all_functions to call the eval functions one by one (by name) and return their results in a dictionary. The method callsite will need to be updated.

Underscores in Python variables are used to indicate private methods so we shouldn't use it to name a public method.

Reply:
Agree that we shouldn't use underscores since it's a public method, and agree that we should aim for readable and maintainable over cleverness.

The best solution here is probably to have `EXCLUDED_FUNCTIONS` as a class variable of the path evals class, rather than a private thing defined anew every time we run `all_functions`. We should also move out the "is this a valid path_eval function" check into a separate function. I'll take a crack at this.